src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -746,11 +758,43 @@ private void FreeLocals(BoundSequence sequence, LocalSymbol doNotRelease)
 
             foreach (var local in sequence.Locals)
             {
-                if ((object)local != doNotRelease)
-                {
-                    FreeLocal(local);
-                }
+                FreeLocal(local);
+            }
+        }
+
+        /// <summary>
+        /// Defines sequence locals and record them so tht they could be retained for the duration of the encompassing expresson
+        /// Use this when taking a reference of the sequence, which can indirectly refer toany of its locals.
+        /// </summary>
+        private void DefineAndRecordLocals(BoundSequence sequence)
+        {
+            if (sequence.Locals.IsEmpty)
+            {
+                return;
+            }
+
+            _builder.OpenLocalScope();
+
+            foreach (var local in sequence.Locals)
+            {
+                var seqLocal = DefineLocal(local, sequence.Syntax);
+                AddExpressionTemp(seqLocal);
+            }
+        }
+
+        /// <summary>
+        /// Closes the visibility/debug scopes for the sequence locals, but keep the local slots from reuse
+        /// for the duration of the encompassing expresson.
+        /// Use this paired with DefineAndRecordLocals when taking a reference of the sequence, which can indirectly refer to any of its locals.
```

Comment:
Scope is in "visibility" sense here - that is used by debugger.- whether "x" is still mapped to a local slot or if it may be resolved to some field.
We keep the actual slots a bit longer though. This is for cases where you no longer can access the locals by name, but may still hold an indirect reference to them. Thus we close visibility scope, but keep the slots.

---
In reply to: [127341301](https://github.com/dotnet/roslyn/pull/20772#discussion_r127341301) [](ancestors = 127341301)